parcels/spatialhash.py

```diff
@@ -242,69 +305,107 @@ def _barycentric_coordinates(nodes, point, min_area=1e-8):
+    # This spreads the bits apart, but introduces overlaps.
+    # Mask 0xff0000ff clears out the unwanted overlaps.
+    n = (n | (n << np.uint32(16))) & np.uint32(0xFF0000FF)
 
-    Parameters
-    ----------
-    lon : np.ndarray
-        2D array of shape (n_y, n_x) containing the longitude of each corner node of the curvilinear grid.
-    lat : np.ndarray
-        2D array of shape (n_y, n_x) containing the latitude of each corner node of the curvilinear grid.
+    # Step 3: Second spreading stage
+    # Similar idea: shift left by 8, OR, then mask.
+    # Now the bits are further separated.
+    n = (n | (n << np.uint32(8))) & np.uint32(0x0300F00F)
 
-    Returns
-    -------
-    xbounds : np.ndarray
-        Array of shape (n_y-1, n_x-1, 2) containing the bounds of each face in the x-direction.
-    ybounds : np.ndarray
-        Array of shape (n_y-1, n_x-1, 2) containing the bounds of each face in the y-direction.
-    """
-    xf = np.stack((lon[:-1, :-1], lon[:-1, 1:], lon[1:, 1:], lon[1:, :-1]), axis=-1)
-    xf_low = xf.min(axis=-1)
-    xf_high = xf.max(axis=-1)
-    xbounds = np.stack([xf_low, xf_high], axis=-1)
+    # Step 4: Third spreading stage
+    # Shift by 4, OR, mask again.
+    # At this point, there are 1 or 2 zeros between many of the bits.
+    n = (n | (n << np.uint32(4))) & np.uint32(0x030C30C3)
+
+    # Step 5: Final spreading stage
+    # Shift by 2, OR, mask.
+    # After this, each original bit is isolated with exactly two zeros
+    # between it and the next bit, ready for 3D Morton interleaving.
+    n = (n | (n << np.uint32(2))) & np.uint32(0x09249249)
 
-    yf = np.stack((lat[:-1, :-1], lat[:-1, 1:], lat[1:, 1:], lat[1:, :-1]), axis=-1)
-    yf_low = yf.min(axis=-1)
-    yf_high = yf.max(axis=-1)
-    ybounds = np.stack([yf_low, yf_high], axis=-1)
+    # Return the dilated value.
+    return n
 
-    return ybounds, xbounds
+
+def _encode_morton3d(x, y, z, xmin, xmax, ymin, ymax, zmin, zmax):
+    """
+    Quantize (x, y, z) to 10 bits each (0..1023), dilate the bits so there are
+    two zeros between successive bits, and interleave them into a 3D Morton code.
+
+    Notes
+    -----
+    - Works with scalars or NumPy arrays (broadcasting applies).
+    - Output is up to 30 bits; we return np.uint32 (or np.uint64 if you prefer).
+    - Requires `part1by2` defined as in your previous snippet.
+    """
+    # Convert inputs to ndarray for consistent dtype/ufunc behavior.
+    x = np.asarray(x)
+    y = np.asarray(y)
+    z = np.asarray(z)
+
+    # --- 1) Normalize each coordinate to [0, 1] over its bounding box. ---
+    # Compute denominators once (avoid division by zero if bounds equal).
+    dx = xmax - xmin
+    dy = ymax - ymin
+    dz = zmax - zmin
+
+    # Normalize to [0,1]; if a range is degenerate, map to 0 to avoid NaN/inf.
+    xn = np.where(dx != 0, (x - xmin) / dx, 0.0)
+    yn = np.where(dy != 0, (y - ymin) / dy, 0.0)
+    zn = np.where(dz != 0, (z - zmin) / dz, 0.0)
```

Comment:
These still give warnings `RuntimeWarning: invalid value encountered in divide` which can be fixed by the following lines

```suggestion
    with np.errstate(invalid="ignore"):
        xn = np.where(dx != 0, (x - xmin) / dx, 0.0)
        yn = np.where(dy != 0, (y - ymin) / dy, 0.0)
        zn = np.where(dz != 0, (z - zmin) / dz, 0.0)
```